{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 508,
   "id": "2e384f79",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data_path_all is the Anndata object containing all cells\n",
    "#data_path_meta is the Anndata object conta \n",
    "#assign_path is the metadata\n",
    "#metacell_types_path gives the cell types of the metacells\n",
    "#benchmarkindices_path imports \n",
    "data_path_all = '/home/labs/shlush/shared/runs_analysis/Metacells_Models/Nili_Human/data_July_2023/all_cells_fil2_500000_cdata_output_with_cell_type_exp_indiv_new.h5ad'\n",
    "data_path_meta = '/home/labs/shlush/shared/runs_analysis/Metacells_Models/Nili_Human/data_July_2023/all_cells_fil2_500000_mdata_output_with_type_new.h5ad'\n",
    "assign_path = '/home/labs/shlush/shared/runs_analysis/Metacells_Models/Nili_Human/cell_metadata'\n",
    "metacell_types_path = '/home/labs/shlush/tomerlev/LOY/metacell_types-2023-03-16.csv'\n",
    "benchmarkindices_path='/home/labs/shlush/noac/workspace/ISSI_2023/indices.csv'\n",
    "all_cells = ad.read_h5ad(data_path_all)\n",
    "metacells = ad.read_h5ad(data_path_meta)\n",
    "benchmarkindices=pd.read_csv(benchmarkindices_path)\n",
    "benchmarkindices=list(benchmarkindices[\"indices\"])\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as colors\n",
    "import seaborn as sns\n",
    "import scipy\n",
    "import statistics\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "9de83139",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Parameters\n",
    "UMIbenchmark = 2000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cfff802",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Find out cell type\n",
    "all_cells.obs.head()\n",
    "cells_to_meta_df = pd.DataFrame()\n",
    "cells_to_meta_df['cnames'] = all_cells.obs['cnames']\n",
    "cells_to_meta_df['metacell'] = all_cells.obs['metacell']\n",
    "cells_to_meta_df.head()\n",
    "\n",
    "metacell_label_df = pd.read_csv(metacell_types_path)\n",
    "meta_type_dict = {}\n",
    "for _, row in metacell_label_df.iterrows():\n",
    "    meta_type_dict[row['metacell']] = row['cell_type']\n",
    "    \n",
    "cell_types = []\n",
    "for _, row in cells_to_meta_df.iterrows():\n",
    "    try:\n",
    "        cell_types.append(meta_type_dict[row['metacell']])\n",
    "    except:\n",
    "        cell_types.append('unassigned')\n",
    "cells_to_meta_df['cell_type'] = cell_types\n",
    "cells_to_meta_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "44a7aa0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#input genes from a chromosome as a list into entries. Make sure to get rid of genes you don't want (in our case it was SLC25A6 and CD99)\n",
    "#input gene \"synonyms\" (e.g. ACOD1 and IRG1)\n",
    "\n",
    "\n",
    "all_cells_genes = all_cells.var_names\n",
    "metacell_genes = metacells.var_names\n",
    "meta_sm = metacells.X\n",
    "ygenes1=['LOC124905300', 'PLCXD1', 'GTPBP6', 'LINC00685', 'PPP2R3B', 'LOC102724521', 'SHOX', 'CRLF2', 'CSF2RA', 'LOC124905238', 'MIR3690', 'IL3RA', 'LOC101928032', 'LOC105373102', 'LINC00106', 'ASMTL-AS1', 'ASMTL', 'P2RY8', 'AKAP17A', 'ASMT', 'LINC02968', 'LOC105379413', 'LOC107985677', 'DHRSX', 'LOC124905239', 'ZBED1', 'LOC101928092', 'MIR6089', 'CD99P1', 'LINC00102', 'XGY2', 'SRY', 'RPS4Y1', 'ZFY', 'ZFY-AS1', 'LINC00278', 'TGIF2LY', 'LOC105377227', 'MIR9985', 'PCDH11Y', 'TTTY23B', 'TSPY2', 'FAM197Y9', 'LINC00280', 'LOC107987353', 'TTTY1B', 'TTTY2B', 'TTTY21B', 'TTTY7', 'TTTY8B', 'AMELY', 'TBL1Y', 'PRKY', 'LOC107987338', 'TTTY16', 'LOC107987339', 'TTTY12', 'LOC105377229', 'LOC105377231', 'LOC105377232', 'LOC107987354', 'LINC00279', 'LOC107987340', 'TTTY18', 'LOC105377234', 'TTTY19', 'LOC105377235', 'TTTY11', 'LOC105379264', 'LOC105379265', 'LOC105379266', 'LOC105379267', 'RBMY1A3P', 'TTTY20', 'TSPY4', 'FAM197Y8', 'TSPY8', 'FAM197Y7', 'FAM197Y6', 'TSPY3', 'FAM197Y5', 'TSPY1', 'FAM197Y4', 'TSPY9', 'FAM197Y3', 'FAM197Y2', 'TSPY10', 'RBMY3AP', 'TTTY8', 'TTTY7B', 'TTTY21', 'TTTY2', 'TTTY1', 'LOC107987342', 'TTTY22', 'TTTY23', 'LOC105379273', 'LOC107987343', 'GYG2P1', 'LOC112268311', 'TTTY15', 'USP9Y', 'DDX3Y', 'LOC112268312', 'UTY', 'MIR12120', 'TMSB4Y', 'VCY', 'VCY1B', 'NLGN4Y', 'NLGN4Y-AS1', 'LOC107987355', 'LOC124900508', 'FAM41AY1', 'LOC105377217', 'FAM224B', 'LOC105377218', 'LOC124905302', 'XKRY', 'CDY2B', 'CDY2A', 'XKRY2', 'LOC124905303', 'LOC105377219', 'FAM224A', 'LOC105377220', 'FAM41AY2', 'LOC124905304', 'HSFY1', 'TTTY9B', 'TTTY9A', 'HSFY2', 'LOC124905305', 'LOC124905307', 'TTTY14', 'LOC124905306', 'BCORP1', 'LOC102723934', 'LOC105377223', 'TXLNGY', 'KDM5D', 'LOC105377224', 'TTTY10', 'EIF1AY', 'RPS4Y2', 'LOC105377225', 'LOC107987346', 'LOC107987347', 'PRORY', 'RBMY2EP', 'RBMY1B', 'RBMY1A1', 'TTTY13', 'LOC105377236', 'LOC105377237', 'RBMY1D', 'RBMY1E', 'LOC107987356', 'PRY2', 'LOC102725532', 'TTTY6B', 'RBMY1F', 'TTTY5', 'RBMY2FP', 'LOC100652931', 'RBMY1J', 'LOC101929148', 'TTTY6', 'PRY', 'LOC107987357', 'LOC107987348', 'TTTY17A', 'TTTY4', 'BPY2', 'LOC105377238', 'DAZ1', 'DAZ2', 'LOC105377239', 'LOC124900506', 'REREP1Y', 'LOC107987349', 'TTTY3B', 'CDY1B', 'GOLGA2P2Y', 'LOC107987350', 'TTTY17B', 'TTTY4B', 'BPY2B', 'LOC105377241', 'DAZ3', 'DAZ4', 'LOC105377242', 'BPY2C', 'TTTY4C', 'TTTY17C', 'SEPTIN14P23', 'GOLGA2P3Y', 'CSPG4P1Y', 'CDY1', 'TTTY3', 'LOC107987352', 'REREP2Y', 'LOC124900507', 'LOC105377244', 'SPRY3', 'VAMP7', 'IL9R', 'WASIR1', 'DDX11L16']\n",
    "ygenes2=['PLCXD1', 'GTPBP6', 'LINC00685', 'PPP2R3B', 'SHOX', 'CRLF2', 'CSF2RA', 'MIR3690', 'IL3RA', 'LINC00106', 'ASMTL-AS1', 'ASMTL', 'P2RY8', 'AKAP17A', 'ASMT', 'DHRSX', 'ZBED1', 'MIR6089', 'CD99P1', 'LINC00102', 'XGY2', 'SRY', 'RPS4Y1', 'ZFY', 'ZFY-AS1', 'LINC00278', 'TGIF2LY', 'MIR9985', 'PCDH11Y', 'TTTY23B', 'TSPY2', 'FAM197Y9', 'LINC00280', 'TTTY1B', 'TTTY2B', 'TTTY21B', 'TTTY7', 'TTTY8B', 'AMELY', 'TBL1Y', 'PRKY', 'TTTY16', 'TTTY12', 'LINC00279', 'TTTY18', 'TTTY19', 'TTTY11', 'RBMY1A3P', 'TTTY20', 'TSPY4', 'FAM197Y8', 'TSPY8', 'FAM197Y7', 'FAM197Y6', 'TSPY3', 'FAM197Y5', 'TSPY1', 'FAM197Y4', 'TSPY9', 'FAM197Y3', 'FAM197Y2', 'TSPY10', 'RBMY3AP', 'TTTY8', 'TTTY7B', 'TTTY21', 'TTTY2', 'TTTY1', 'TTTY22', 'TTTY23', 'GYG2P1', 'TTTY15', 'USP9Y', 'DDX3Y', 'UTY', 'MIR12120', 'TMSB4Y', 'VCY', 'VCY1B', 'NLGN4Y', 'NLGN4Y-AS1', 'FAM41AY1', 'FAM224B', 'XKRY', 'CDY2B', 'CDY2A', 'XKRY2', 'FAM224A', 'FAM41AY2', 'HSFY1', 'TTTY9B', 'TTTY9A', 'HSFY2', 'TTTY14', 'BCORP1', 'TXLNGY', 'KDM5D', 'TTTY10', 'EIF1AY', 'RPS4Y2', 'PRORY', 'RBMY2EP', 'RBMY1B', 'RBMY1A1', 'TTTY13', 'RBMY1D', 'RBMY1E', 'PRY2', 'TTTY6B', 'RBMY1F', 'TTTY5', 'RBMY2FP', 'LOC100652931', 'RBMY1J', 'LOC101929148', 'TTTY6', 'PRY', 'TTTY17A', 'TTTY4', 'BPY2', 'DAZ1', 'DAZ2', 'TTTY3B', 'CDY1B', 'GOLGA2P2Y', 'TTTY17B', 'TTTY4B', 'BPY2B', 'DAZ3', 'DAZ4', 'BPY2C', 'TTTY4C', 'TTTY17C', 'SEPTIN14P23', 'GOLGA2P3Y', 'CSPG4P1Y', 'CDY1', 'TTTY3', 'SPRY3', 'VAMP7', 'IL9R', 'WASIR1', 'DDX11L16']\n",
    "entries=(list(set(ygenes1)|set(ygenes2)))\n",
    "\n",
    "synonyms = [\"TSPY9P\"]\n",
    "og_chry_gene_names = set(entries)\n",
    "\n",
    "\n",
    "not_included = set([])\n",
    "included = set([])\n",
    "\n",
    "#detect if chromsome genes are in the anndata gene list\n",
    "#chry_gene_names is the y chromsome gene names in the anndata gene list\n",
    "#subset_df is the dataframe containing gene expression of Y chromosome genes\n",
    "for gene in og_chry_gene_names:\n",
    "    if gene in all_cells_genes:\n",
    "        included.add(gene)  # if y-chr gene, put in included\n",
    "    else: not_included.add(gene) # else, not_included -- doesn't add not included synonyms\n",
    "\n",
    "for gene in synonyms:\n",
    "    if gene in all_cells_genes:\n",
    "        included.add(gene)\n",
    "\n",
    "chry_gene_names = included \n",
    "\n",
    "subset = all_cells[all_cells.obs_names,list(chry_gene_names)] # gives subset of anndata object including only \"included\" y-cgenes\n",
    "\n",
    "subset_df = pd.DataFrame.sparse.from_spmatrix(subset.X, columns=subset.var_names)\n",
    "subset_df.index = subset.obs_names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "5aeab474",
   "metadata": {},
   "outputs": [],
   "source": [
    "#creating a DF with UMI and total UMi for each Y gene from every cell\n",
    "cells_counts_df = subset_df\n",
    "\n",
    "#adding all UMI's per Y gene, concat to different dataframe\n",
    "umi_sum=list(cells_counts_df.sum())\n",
    "umi_sum=pd.DataFrame(umi_sum)\n",
    "umi_sum=umi_sum.transpose()\n",
    "\n",
    "\n",
    "umi_sum.columns=cells_counts_df.columns\n",
    "cells_counts_df=pd.concat([cells_counts_df,umi_sum])\n",
    "new_names={0:\"Total UMI\"}\n",
    "cells_counts_df=cells_counts_df.rename(index=new_names)\n",
    "\n",
    "temp_df=subset_df\n",
    "cells_expressed=[(temp_df[col]>0.0).sum() for col in temp_df.columns]\n",
    "cells_expressed=pd.DataFrame(cells_expressed)\n",
    "cells_expressed=cells_expressed.transpose()\n",
    "cells_expressed.columns=cells_counts_df.columns\n",
    "cells_counts_df=pd.concat([cells_counts_df,cells_expressed])\n",
    "\n",
    "new_names={0:\"Number of Cells Expressing Gene\"}\n",
    "cells_counts_df=cells_counts_df.rename(index=new_names)\n",
    "#cells_counts_df contains the gene expression data for all Y genes and has two additional rows, Total UMI of y genes and number of cells expressing the y gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "1cdd954a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#creating a DF (alldf1) with gene expression and total UMI for each gene from every cell\n",
    "all_cells\n",
    "alldf=pd.DataFrame.sparse.from_spmatrix(all_cells.X, columns=all_cells.var_names)\n",
    "alldf.index = subset.obs_names\n",
    "alldf1 = alldf\n",
    "umi_sum1=list(alldf1.sum())\n",
    "umi_sum1=pd.DataFrame(umi_sum1)\n",
    "umi_sum1=umi_sum1.transpose()\n",
    "\n",
    "umi_sum1.columns=alldf1.columns\n",
    "alldf1=pd.concat([alldf1,umi_sum1])\n",
    "new_names={0:\"Total UMI\"}\n",
    "alldf1=alldf1.rename(index=new_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "79225415",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sexlist is list of sex for every cell, agelist is list for every age\n",
    "#ydf is cells_counts_df but a column of Sex and Age for cells is added\n",
    "#alldf1 is alldf but with Sex and Age with only Gene Expression\n",
    "ydf=cells_counts_df\n",
    "metadata_df = pd.read_csv(assign_path, sep=\" \", header=0)\n",
    "samplesex=metadata_df['sex']\n",
    "sexlist=samplesex.to_list()\n",
    "sexlist.append(\"Total UMI\")\n",
    "sexlist.append(\"Number of Cells Expressing Gene\")\n",
    "agelist=metadata_df['age'].to_list()\n",
    "agelist.append(\"Total UMI\")\n",
    "agelist.append(\"Number of Cells Expressing Gene\")\n",
    "ydf[\"Sex\"]=sexlist\n",
    "ydf[\"Age\"]=agelist\n",
    "mydf=ydf[ydf[\"Sex\"]==\"male\"]\n",
    "fydf=ydf[ydf[\"Sex\"]==\"female\"]\n",
    "nydf=ydf[ydf[\"Sex\"].isna()]\n",
    "len(mydf.index.to_list())+len(fydf.index.to_list())+len(nydf.index.to_list())\n",
    "sexlist1=sexlist\n",
    "sexlist1.remove('Number of Cells Expressing Gene')\n",
    "\n",
    "agelist1=agelist\n",
    "agelist1.remove('Number of Cells Expressing Gene')\n",
    "alldf1['Sex']=sexlist1\n",
    "alldf1['Age']=agelist1\n",
    "tempdf1=alldf1\n",
    "tempdf1.drop('Sex', axis=1,inplace=True)\n",
    "tempdf1.drop('Age', axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5017584",
   "metadata": {},
   "outputs": [],
   "source": [
    "#list of total umis per cell\n",
    "totalumis=scipy.sparse.csr_matrix.sum(all_cells.X,axis=1)\n",
    "\n",
    "totalumis=totalumis.tolist()\n",
    "\n",
    "umisum=[]\n",
    "for i in range (0,360810):\n",
    "    umisum.append(totalumis[i][0])\n",
    "    \n",
    "tempydf=ydf\n",
    "yumisum=list(tempydf.sum(axis=1))\n",
    "\n",
    "yumisum.pop()\n",
    "yumisum.pop()\n",
    "\n",
    "#Calculating Y Score\n",
    "yscorelist=[]\n",
    "for i in range (0,360810):\n",
    "    if totalumis[i][0]==0:\n",
    "        yscorelist.append(0)\n",
    "    else:\n",
    "        yscorelist.append(yumisum[i]/totalumis[i][0])\n",
    "        \n",
    "cellids=tempdf1.index.tolist()\n",
    "cellids.pop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "7f5165c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#return indices of cells to keep after filtering by UMI benchmark for every cell in keepindices\n",
    "keepindices=[]\n",
    "for i in range(0,len(totalumis)):\n",
    "    if totalumis[i][0]>UMIbenchmark:\n",
    "         keepindices.append(i)\n",
    "\n",
    "#returns list of yscores of cells after filter in keepyscorelist\n",
    "#returns list of cellids of cells after filter in keepcellids\n",
    "keepyscorelist=[]\n",
    "keepcellids=[]\n",
    "for i in keepindices:\n",
    "    keepyscorelist.append(yscorelist[i])\n",
    "    keepcellids.append(cellids[i])\n",
    "\n",
    "#yscoredf: Dataframe with each filtered cell and its Y score\n",
    "yscoredf=pd.DataFrame(keepcellids,columns=[\"Cell ID\"])\n",
    "yscoredf['Y Score']=keepyscorelist\n",
    "\n",
    "data = {\n",
    "    'Cell ID': cellids,\n",
    "    'Y Score': yscorelist\n",
    "}\n",
    "\n",
    "yscoredf = pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "7b826df9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#return indices of male cells, female cells, and cells with unknown sex\n",
    "maleindices=[]\n",
    "femaleindices=[]\n",
    "naindices=[]\n",
    "for i in range (0,360810):\n",
    "    if ydf[\"Sex\"][i]==\"male\":\n",
    "        maleindices.append(i)\n",
    "    elif ydf[\"Sex\"][i]==\"female\":\n",
    "        femaleindices.append(i)\n",
    "    else:\n",
    "        naindices.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "id": "4a0366f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "yscorelist=yscoredf['Y Score'].tolist()\n",
    "LOYindex=[]\n",
    "notLOYindex=[]\n",
    "#returns indices of LOY and indices of not LOY cells\n",
    "for i in range(0, len(yscoredf.index)):\n",
    "    if yscorelist[i]<0.001:\n",
    "        LOYindex.append(i)\n",
    "    else:\n",
    "        notLOYindex.append(i)\n",
    "\n",
    "#Removes all female cells from the LOY and not LOY index lists and removes cells that don't pass the benchmark\n",
    "LOYindex=list(set(LOYindex)&set(maleindices)&set(benchmarkindices))\n",
    "notLOYindex=list(set(notLOYindex)&set(maleindices)&set(benchmarkindices))\n",
    "#creates sparce matrices of male cells that pass the benchmark with LOY or not LOY\n",
    "LOYsparse=all_cells[all_cells.obs_names[LOYindex],all_cells_genes]\n",
    "notLOYsparse=all_cells[all_cells.obs_names[notLOYindex],all_cells_genes]\n",
    "#turns said sparce matrices into dfs\n",
    "LOYdf=pd.DataFrame.sparse.from_spmatrix(LOYsparse.X, columns=LOYsparse.var_names)\n",
    "notLOYdf=pd.DataFrame.sparse.from_spmatrix(notLOYsparse.X, columns=notLOYsparse.var_names)\n",
    "#returns umi totals for genes amongst LOY and not LOY cell groups\n",
    "LOYdfsum=LOYdf.sum()\n",
    "notLOYdfsum=notLOYdf.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf94b6d7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#sum the total umi for LOY cells and total umi for not LOY cells. Used for normalization\n",
    "LOYsum=sum([umisum[i] for i in range(0,len(umisum)) if i in LOYindex])\n",
    "noLOYsum=sum([umisum[i] for i in range(0,len(umisum)) if i in noLOYindex])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e92d0da",
   "metadata": {},
   "outputs": [],
   "source": [
    "#graph of male vs female Y scores\n",
    "fig, ax = plt.subplots(figsize=(12,9))\n",
    "sns.kdeplot(np.array(yscorelist)[list(set(maleindices)&set(benchmarkindices))], shade=True, color='blue', label='Male', ax=ax)\n",
    "sns.kdeplot(np.array(yscorelist)[list(set(femaleindices)&set(benchmarkindices))], shade=True, color='red', label='Female', ax=ax)\n",
    "ax.legend()\n",
    "ax.set_title('Male vs Female Cells Y-Scores')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "id": "32bee6b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identify 0s\n",
    "zeroindex=[i for i in range(0,len(umisum)) if (umisum[i]==0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c42ff88",
   "metadata": {},
   "outputs": [],
   "source": [
    "#umisum is sum of umis per gene. Log normalizes\n",
    "labels=ydf.columns\n",
    "logLOY=[math.log(LOYumi[i]/LOYsum) for i in range(0,len(LOYumi))]\n",
    "lognoLOY=[math.log(noLOYumi[i]/noLOYsum) for i in range(0,len(noLOYumi))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "id": "4b66d975",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identify genes with 0 UMIs in LOY or no LOY \n",
    "zeroLOY=[i for i in range(0, 33487) if LOYdfsum[i]==0]\n",
    "zeronoLOY=[i for i in range(0, 33487) if notLOYdfsum[i]==0]\n",
    "zeros=list(set(zeroLOY)|set(zeronoLOY))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "id": "ad3faf9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#identify indices of y genes to be removed\n",
    "yindices=[]\n",
    "for i in range(0,33487):\n",
    "    if all_cells_genes[i] in chry_gene_names:\n",
    "        yindices.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfb99829",
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove index of genes with 0 total umi or are y genes. returns genes that have gene umi>0 and aren't y genes\n",
    "notLOYdfindexlist=list(range(0,33487))\n",
    "for i in list(set(zeros)|set(yindices)): \n",
    "    print(i)\n",
    "    notLOYdfindexlist.remove(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "753e0891",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Non log transformed plot without 0 UMI genes of UMI differential gene expression between LOY and non LOY\n",
    "fig, ax = plt.subplots(figsize=(20,20))\n",
    "sns.scatterplot(x=(notLOYdf.sum()[notLOYdfindexlist])/noLOYsumy, y=LOYdf.sum()[notLOYdfindexlist]/LOYsumy,color=\"blue\") # Add labels to each point \n",
    "sns.regplot(x=notLOYdf.sum()[notLOYdfindexlist]/noLOYsumy, y=LOYdf.sum()[notLOYdfindexlist]/LOYsumy)\n",
    "plt.xlabel('No LOY Gene Total UMI') \n",
    "plt.ylabel('LOY Gene Total UMI') \n",
    "plt.title('Removed 0 UMI LOY vs No LOY Gene Expression in Male Cells Benchmark 2000')\n",
    "plt.savefig('LOY vs No LOY Gene Expression in Male Cells Benchmark 2000.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "id": "1d31c90e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#log transform LOY\n",
    "lognotLOY=[]\n",
    "for i in list(notLOYdf.sum()[notLOYdfindexlist]/noLOYsumy):\n",
    "    lognotLOY.append(math.log(i))\n",
    "logLOY=[]\n",
    "for i in list(LOYdf.sum()[notLOYdfindexlist]/LOYsumy):\n",
    "    logLOY.append(math.log(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faa379f1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#log transformed UMI LOY vs no LOY\n",
    "fig, ax = plt.subplots(figsize=(20,20))\n",
    "sns.scatterplot(x=np.array(lognotLOY), y=np.array(logLOY),color=\"blue\") # Add labels to each point \n",
    "sns.regplot(x=np.array(lognotLOY), y=np.array(logLOY))\n",
    "label=list(all_cells_genes[notLOYdfindexlist])\n",
    "plt.xlabel('No LOY Gene Total UMI') \n",
    "plt.ylabel('LOY Gene Total UMI') \n",
    "plt.title('Log Transformed LOY vs No LOY Gene Expression in Male Cells Benchmark 2000')\n",
    "plt.savefig('Log LOY vs No LOY Gene Expression in Male Cells Benchmark 2000.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7b5b256",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Residual plot of log transformed UMIs\n",
    "fig, ax = plt.subplots(figsize=(50,50))\n",
    "model = sns.regplot(x=lognotLOY, y=logLOY, scatter=True, fit_reg=True)\n",
    "x_regression = model.get_lines()[0].get_xdata()\n",
    "y_regression = model.get_lines()[0].get_ydata()\n",
    "plt.clf()\n",
    "# Calculate residuals\n",
    "residuals = np.array(logLOY) - np.interp(lognotLOY, x_regression, y_regression)\n",
    "\n",
    "sns.residplot(x=np.array(lognotLOY), y=residuals, lowess=True, line_kws={'color': 'red'})\n",
    "\n",
    "# # Adding labels and title\n",
    "plt.xlabel(\"No LOY Gene Total UMI\")\n",
    "plt.ylabel(\"LOY Gene Total UMI Residuals\")\n",
    "label=list(all_cells_genes[notLOYdfindexlist])\n",
    "for i, residual in enumerate(residuals):\n",
    "    if abs(residual) > 0.5:\n",
    "        plt.annotate(label[i], (np.array(lognotLOY)[i], residual))\n",
    "\n",
    "plt.title('Large Residual Plot Log Transformed LOY vs No LOY Gene Expression in Male Cells Benchmark 2000')\n",
    "plt.savefig('Large Residual Plot Log LOY vs No LOY Gene Expression in Male Cells Benchmark 2000.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "19e839d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get unique cell type names, get list of cell types for all cells\n",
    "celltypelist=list(np.unique(list(all_cells.obs[\"cell_type\"])))\n",
    "celltypelonglist=list(all_cells.obs[\"cell_type\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "db935ede",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get indices for every cell type\n",
    "celltypeindex={}\n",
    "for i in celltypelist:\n",
    "    templist=[]\n",
    "    for j in range(0,len(all_cells.obs.index)):\n",
    "        if celltypelonglist[j]==i:\n",
    "            templist.append(j)\n",
    "    celltypeindex[i]=templist\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fd647bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate graphs of Total Umi of Genes in LOY vs not LOY groups for regular UMI, log transformed UMI, and residuals. \n",
    "\n",
    "yscorelist=yscoredf['Y Score'].tolist()\n",
    "LOYindex=[]\n",
    "notLOYindex=[]\n",
    "\n",
    "for celltype in celltypelist:\n",
    "    for i in range(0, len(yscoredf.index)):\n",
    "        if yscorelist[i]<0.001:\n",
    "            LOYindex.append(i)\n",
    "        else:\n",
    "            notLOYindex.append(i)\n",
    "    LOYindex=list(set(LOYindex)&set(maleindices)&set(benchmarkindices)&set(celltypeindex[celltype]))\n",
    "    notLOYindex=list(set(notLOYindex)&set(maleindices)&set(benchmarkindices)&set(celltypeindex[celltype]))\n",
    "    LOYsparse=all_cells[all_cells.obs_names[LOYindex],all_cells_genes]\n",
    "    notLOYsparse=all_cells[all_cells.obs_names[notLOYindex],all_cells_genes]\n",
    "    LOYdf=pd.DataFrame.sparse.from_spmatrix(LOYsparse.X, columns=LOYsparse.var_names)\n",
    "    notLOYdf=pd.DataFrame.sparse.from_spmatrix(notLOYsparse.X, columns=notLOYsparse.var_names)\n",
    "    LOYdfsum=LOYdf.sum()\n",
    "    notLOYdfsum=notLOYdf.sum()\n",
    "    print(LOYindex)\n",
    "    print(notLOYindex)\n",
    "    zeroLOY=[i for i in range(0, 33487) if LOYdfsum[i]==0]\n",
    "    zeronoLOY=[i for i in range(0, 33487) if notLOYdfsum[i]==0]\n",
    "    zeros=list(set(zeroLOY)|set(zeronoLOY))\n",
    "    \n",
    "    notLOYdfindexlist=list(range(0,33487))\n",
    "    for i in list(set(zeros)|set(yindices)): \n",
    "        notLOYdfindexlist.remove(i)\n",
    "    print(celltype)\n",
    "    if (notLOYdfindexlist!=[]):\n",
    "        fig, ax = plt.subplots(figsize=(20,20))\n",
    "        print(notLOYdfindexlist)\n",
    "        print(notLOYdf.sum()[notLOYdfindexlist])\n",
    "        sns.scatterplot(x=(notLOYdf.sum()[notLOYdfindexlist])/noLOYsumy, y=LOYdf.sum()[notLOYdfindexlist]/LOYsumy,color=\"blue\") # Add labels to each point \n",
    "        sns.regplot(x=notLOYdf.sum()[notLOYdfindexlist]/noLOYsumy, y=LOYdf.sum()[notLOYdfindexlist]/LOYsumy)\n",
    "        plt.xlabel('No LOY Gene Total UMI') \n",
    "        plt.ylabel('LOY Gene Total UMI') \n",
    "        plt.title(f'{celltype} Removed 0 UMI LOY vs No LOY Gene Expression in Male Cells Benchmark 2000')\n",
    "        plt.savefig(f'{celltype} Removed 0 LOY vs No LOY Gene Expression in Male Cells Benchmark 2000.png')\n",
    "        plt.show()\n",
    "\n",
    "        lognotLOY=[]\n",
    "        for i in list(notLOYdf.sum()[notLOYdfindexlist]/noLOYsumy):\n",
    "            lognotLOY.append(math.log(i))\n",
    "\n",
    "        logLOY=[]\n",
    "        for i in list(LOYdf.sum()[notLOYdfindexlist]/LOYsumy):\n",
    "            logLOY.append(math.log(i))\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(20,20))\n",
    "        sns.scatterplot(x=np.array(lognotLOY), y=np.array(logLOY),color=\"blue\") # Add labels to each point \n",
    "        sns.regplot(x=np.array(lognotLOY), y=np.array(logLOY))\n",
    "        label=list(all_cells_genes[notLOYdfindexlist])\n",
    "        plt.xlabel('No LOY Gene Total UMI') \n",
    "        plt.ylabel('LOY Gene Total UMI') \n",
    "        plt.title(f'{celltype} Log Transformed LOY vs No LOY Gene Expression in Male Cells Benchmark 2000')\n",
    "        plt.savefig(f'{celltype} Log LOY vs No LOY Gene Expression in Male Cells Benchmark 2000.png')\n",
    "        plt.show()\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(30,30))\n",
    "        model = sns.regplot(x=lognotLOY, y=logLOY, scatter=True, fit_reg=True)\n",
    "        x_regression = model.get_lines()[0].get_xdata()\n",
    "        y_regression = model.get_lines()[0].get_ydata()\n",
    "        plt.clf()\n",
    "        # Calculate residuals\n",
    "        residuals = np.array(logLOY) - np.interp(lognotLOY, x_regression, y_regression)\n",
    "\n",
    "        sns.residplot(x=np.array(lognotLOY), y=residuals, lowess=True, line_kws={'color': 'red'})\n",
    "        plt.xlabel(\"No LOY Gene Total UMI\")\n",
    "        plt.ylabel(\"LOY Gene Total UMI Residuals\")\n",
    "        label=list(all_cells_genes[notLOYdfindexlist])\n",
    "        for i, residual in enumerate(residuals):\n",
    "            if abs(residual) > 0.5:\n",
    "                plt.annotate(label[i], (np.array(lognotLOY)[i], residual))\n",
    "\n",
    "        plt.title(f'{celltype} Large Residual Plot Log Transformed LOY vs No LOY Gene Expression in Male Cells Benchmark 2000')\n",
    "        plt.savefig(f'{celltype} Large Residual Plot Log LOY vs No LOY Gene Expression in Male Cells Benchmark 2000.png')\n",
    "        plt.show()\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(30,30))\n",
    "        model = sns.regplot(x=lognotLOY, y=logLOY, scatter=True, fit_reg=True)\n",
    "        x_regression = model.get_lines()[0].get_xdata()\n",
    "        y_regression = model.get_lines()[0].get_ydata()\n",
    "        plt.clf()\n",
    "        residuals = np.array(logLOY) - np.interp(lognotLOY, x_regression, y_regression)\n",
    "\n",
    "        sns.residplot(x=np.array(lognotLOY), y=residuals, lowess=True, line_kws={'color': 'red'})\n",
    "        plt.xlabel(\"No LOY Gene Total UMI\")\n",
    "        plt.ylabel(\"LOY Gene Total UMI Residuals\")\n",
    "        label=list(all_cells_genes[notLOYdfindexlist])\n",
    "        for i, residual in enumerate(residuals):\n",
    "            if abs(residual) > 0.5:\n",
    "                plt.annotate(label[i], (np.array(lognotLOY)[i], residual))\n",
    "\n",
    "        plt.title(f'{celltype} Large Residual Plot Log Transformed LOY vs No LOY Gene Expression in Male Cells Benchmark 2000')\n",
    "        plt.savefig(f'{celltype} Large Residual Plot Log LOY vs No LOY Gene Expression in Male Cells Benchmark 2000.png')\n",
    "        plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
